### PR TITLE
`azurerm_key_vault_access_policy` - check access policy consistency before committing to state

### LINKED ISSUE #9107: "Unauthorized" error when creating `azurerm_key_vault_secret` while checking for secret existence in v2.34
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

[`v2.34.0`](https://github.com/terraform-providers/terraform-provider-azurerm/releases/tag/v2.34.0)

### Affected Resource(s)

* `azurerm_key_vault_secret`

### Terraform Configuration Files

```hcl
provider "azurerm" {
  subscription_id = var.ARM_SUBSCRIPTION_ID
  tenant_id       = var.ARM_TENANT_ID
  features {}
}

terraform {
  backend "azurerm" {
  }
}

resource "azurerm_resource_group" "business_code" {
  name     = "${var.bc_id}-rg"
  location = "westeurope"
}

module "keyvault" {
  source                   = "./modules/keyvault"
  keyvault_name            = "${var.bc_id}-kv"
  resource_group_name      = azurerm_resource_group.business_code.name
  location                 = azurerm_resource_group.business_code.location
  tenant_id                = var.ARM_TENANT_ID
  object_ids               = ["728b6df8-1aa8-11eb-a463-13524f6265db"]
  soft_delete_enabled      = false
  purge_protection_enabled = false
}

resource "random_string" "values" {
  count  = 30
  length = 16
}

resource "azurerm_key_vault_secret" "bc_secrets" {
  count        = 30
  name         = "secret${count.index}"
  value        = random_string.values[count.index].result
  key_vault_id = module.keyvault.id
}
```

### Debug Output

```
2020-10-30T11:23:19.179+0100 [DEBUG] plugin.terraform-provider-azurerm_v2.34.0_x5: {"error":{"code":"Unauthorized","message":"Request is missing a Bearer or PoP token."}}
```

Longer version: [error for secret8 (Unauthorized)](https://gist.github.com/bastiandg/7dd0d00db3fe71108b48e91cc0d13ed4)

### Expected Behavior

A keyvault secret is created.

### Actual Behavior

```
Error: Error checking for presence of existing Secret "secret8" (Key Vault "https://bgtf2-kv.vault.azure.net/"): keyvault.BaseClient#GetSecret: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=5ec7e6b0-1aa9-11eb-8425-efaf4d203e40;oid=5ec7e6b0-1aa9-11eb-8425-efaf4d203e40;numgroups=17;iss=https://sts.windows.net/5ec7e6b0-1aa9-11eb-8425-efaf4d203e40/' does not have secrets get permission on key vault 'bgtf2-kv;location=westeurope'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"AccessDenied"}

  on main.tf line 33, in resource "azurerm_key_vault_secret" "bc_secrets":
  33: resource "azurerm_key_vault_secret" "bc_secrets" {
```

### Steps to Reproduce

1. `terraform init`
1. `terraform apply`

### Important Factoids

The same configuration works flawlessly with version `2.33.0`.

### PR DESCRIPTION
**Pull request by jackofallops**
fixes #9107

### CODE REVIEW

File: azurerm/internal/services/keyvault/key_vault_access_policy_resource.go [hunk 1]
```diff
@@ -327,3 +351,25 @@ func FindKeyVaultAccessPolicy(policies *[]keyvault.AccessPolicyEntry, objectId s
...
+
+		read, err := client.Get(ctx, resourceGroup, vaultName)
+		if err != nil {
+			if utils.ResponseWasNotFound(read.Response) {
+				return "vaultnotfound", "vaultnotfound", fmt.Errorf("failed to find vault %q (resource group %q)", vaultName, resourceGroup)
+			}
+		}
+
+		if read.Properties != nil && read.Properties.AccessPolicies != nil {
+			policy, _ := FindKeyVaultAccessPolicy(read.Properties.AccessPolicies, objectId, applicationId)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/key_vault_access_policy_resource.go:367]: this should still check for the error here

File: azurerm/internal/services/keyvault/key_vault_access_policy_resource.go [hunk 2]
```diff
@@ -203,6 +205,28 @@ func resourceArmKeyVaultAccessPolicyCreateOrDelete(d *schema.ResourceData, meta
...
+		stateConf.Timeout = d.Timeout(schema.TimeoutDelete)
+	}
+
+	if action == keyvault.Replace {
+		stateConf.Timeout = d.Timeout(schema.TimeoutUpdate)
+	}
+
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("failed waiting for Key Vault Access Policy (Object ID: %q) to apply: %+v", objectId, err)
+	}
```
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/key_vault_access_policy_resource.go:229]: (as discussed offline) this likely wants adding to the key vault resource's update function, too - to ensure that's covered from this edge-case too?

**Review by tombuildsstuff — APPROVED** [review]
otherwise LGTM 👍

File: azurerm/internal/services/keyvault/key_vault_access_policy_resource.go (hunk 1 above)
Comment (jackofallops) [reply azurerm/internal/services/keyvault/key_vault_access_policy_resource.go:367]: err is always `nil` for `FindKeyVaultAccessPolicy`, not picked up by the linter. Fix the func?

**Comment by jackofallops** [comment]
Tests passing:
image

**Pull request merged**